src/apps/components/juttle-viewer.js

```diff
@@ -1,9 +1,54 @@
 import React from 'react';
 
-const JuttleViewer = (props) => {
-    return (
-        <pre className='juttle-source'>{props.bundle.program}</pre>
-    );
+import JuttleEditor from './juttle-editor';
+import { newBundle } from '../actions';
+import { LOCAL_BUNDLE_ID } from '../constants';
+
+class JuttleViewer extends React.Component {
+    constructor(props) {
+        super(props);
+        let isNewBundle = !props.bundle;
+        if (isNewBundle) {
+            props.dispatch(newBundle(LOCAL_BUNDLE_ID, {program: ''}, {}));
+        }
+
+        let query = window.location.search || '';
+        let isLocal = query.match(/\?local/);
+        this.state = {
+            readOnly: !isNewBundle && !isLocal
+        };
+    }
+
+    edit() {
+        window.history.pushState('local', 'local', '?local');
+
+        if (this.props.bundle) {
+            localStorage.setItem('program', this.props.bundle.program);
+        }
+
+        this.setState({
+            readOnly: false
+        });
+    }
+
+    render() {
+        let editButton = this.state.readOnly ? <button
+            onClick={this.edit.bind(this)}
+            className="btn btn-default btn-code btn-edit">
+            edit
+        </button> : null;
+        return (
+            <div className="juttle-editor">
+                {editButton}
```

Comment:
lines 15-42 are what we use [react-router](https://github.com/reactjs/react-router) for.

It can be written as such:
```jsx
handleClick() {
    if (this.props.bundle) {
        localStorage.setItem('program', this.props.bundle.program);
    }
}
 
render() {
    return (
        <Link to={{ pathname: '/', query: { local: true }}} onClick={this.handleClick} className="btn btn-default" />
    );
}
```